docs/design/indexed-job.md

```diff
@@ -0,0 +1,891 @@
+complete.
+
+Parameter lists could be passed inside an EnvVar.  This would have length
+limitations, would pollute the output of `kubectl describe pods` and `kubectl
+get pods -o json`.
+
+Parameter lists could be passed inside an annotation.  This would have length
+limitations, would pollute the output of `kubectl describe pods` and `kubectl
+get pods -o json`.  Also, currently annotations can only be extracted into a
+single file.  Complex logic is then needed to filter out exactly the desired
+annotation data.
+
+Bash array variables could simplify extraction of a particular parameter from a
+list of parameters.  However, some popular base images do not include
+`/bin/bash`.  For example, `busybox` uses a compact `/bin/sh` implementation
+that does not support array syntax.
+
+Kubelet does support [expanding varaibles without a
+shell](http://kubernetes.io/v1.1/docs/design/expansion.html).  But it does not
+allow for recursive substitution, which is required to extract the correct
+parameter from a list based on the completion index of the pod.  The syntax
+could be extended, but doing so seems complex and will be an unfamiliar syntax
+for users.
+
+Putting all the command line editing into a script and running that causes
+the least pollution to the original command line, and it allows
+for complex error handling.
+
+Kubectl could store the script in an [Inline Volume](
+https://github.com/kubernetes/kubernetes/issues/13610) if that proposal
+is approved. That would remove the need to manage the lifetime of the
+configData/secret, and prevent the case where someone changes the
+configData mid-job, and breaks things in a hard-to-debug way.
+
+
+### Comparison of the two options
+
+
+Index-only approach relies on image containing certain shell commands, which
+are not guaranteed to be available on all images.  Kubectl cannot warn user at
+time of running kubectl on whether image has needed tools.
+
+Multiple-substitution option has long experience inside Google, and while not perfect
+is a devil-some-of-us-know.
+
+Multiple-substitution maps in a known way to mixing configuration and programming
+to generate config.  Using the index-only approach from
+non-kubectl clients requires that they mimic the script-generation step,
+which is pretty gross.  Or that all programs run in an indexed job are able
+to be parameterized only by a single integer variable, the index.
+
+## Open Issues
+
+For Workqueue Jobs, completions has no meaning.  Parallelism should be allowed to be greater than it, and pods have no identity.  So, the job controller should not create a scoreboard in the JobStatus, just a count.  Therefore, we need to add one of the following to JobSpec:
+
+- allow unset `.spec.completions` to indicate no scoreboard, and no index for tasks (identical tasks)
+- allow `.spec.completions=-1` to indicate the same.
+- add `.spec.indexed` to job to indicate need for scoreboard.
```

Comment:
In any of the above case, how the controller will know when the job is finished?